File: app/src/main/java/com/yixiongsun/particlelivebackground/BokehRainbowCircle.java
package com.yixiongsun.particlelivebackground;

import android.graphics.Bitmap;

public class BokehRainbowCircle {
    float origRadius;

    float deltaRadius;

    float radius;

    float origX;

    float deltaX;

    float x;

    float origY;

    float deltaY;

    float y;

    int color;

    int alpha;

    int steps;

    int currentStep;

    Bitmap bitmap;

    public BokehRainbowCircle(float xCenter, float yCenter, float radius,
                              int color, int steps) {
        this.x = xCenter;
        this.origX = xCenter;
        this.deltaX = (float) (40.0 * Math.random() - 20.0);

        this.y = yCenter;
        this.origY = yCenter;
        this.deltaY = (float) (40.0 * Math.random() - 20.0);

        this.origRadius = radius;
        this.radius = radius;
        this.deltaRadius = 0.5f * radius;

        this.color = color;
        this.alpha = 0;

        this.steps = steps;
    }

    void tick() {
        this.currentStep++;

        float fraction = (float) this.currentStep / (float) this.steps;

        this.radius = this.origRadius + fraction * this.deltaRadius;
        this.x = this.origX + fraction * this.deltaX;
        this.y = this.origY + fraction * this.deltaY;

        if (fraction <= 0.25f) {
            this.alpha = (int) (128 * 4.0f * fraction);
        } else {
            this.alpha = (int) (-128 * (fraction - 1) / 0.75f);
        }
    }

    boolean isDone() {
        return this.currentStep > this.steps;
    }
}

File: app/src/main/java/com/yixiongsun/particlelivebackground/BokehRainbowCircle.kt
package com.yixiongsun.particlelivebackground

import android.graphics.Bitmap

class BokehRainbowCircle(
    internal var x: Float, internal var y: Float, internal var origRadius: Float,
    internal var color: Int, internal var steps: Int
) {

    internal var deltaRadius: Float = 0.toFloat()

    internal var radius: Float = 0.toFloat()

    internal var origX: Float = 0.toFloat()

    internal var deltaX: Float = 0.toFloat()

    internal var origY: Float = 0.toFloat()

    internal var deltaY: Float = 0.toFloat()

    internal var alpha: Int = 0

    internal var currentStep: Int = 0

    internal var bitmap: Bitmap? = null

    internal val isDone: Boolean
        get() = this.currentStep > this.steps

    init {
        this.origX = x
        this.deltaX = (40.0 * Math.random() - 20.0).toFloat()
        this.origY = y
        this.deltaY = (40.0 * Math.random() - 20.0).toFloat()
        this.radius = origRadius
        this.deltaRadius = 0.5f * origRadius
        this.alpha = 0
    }

    internal fun tick() {
        this.currentStep++

        val fraction = this.currentStep.toFloat() / this.steps.toFloat()

        this.radius = this.origRadius + fraction * this.deltaRadius
        this.x = this.origX + fraction * this.deltaX
        this.y = this.origY + fraction * this.deltaY

        if (fraction <= 0.25f) {
            this.alpha = (128f * 4.0f * fraction).toInt()
        } else {
            this.alpha = (-128 * (fraction - 1) / 0.75f).toInt()
        }
    }
}